File: swift_algorithm/data_structure/queue.swift
#!/usr/bin/swift


/**
 Queue也使用Array, enqueue为O(1), dequeue为O(n)(因dequeue移出首元素, 剩下的元素需要依次向前做shift移动操作).
 因swift中的Array通常使用顺序表, append的平均复杂度为O(1). 个别情况考虑顺序表的分配内存已满而需要的迁移策略.
 迁移涉及到分配新的内存空间, 将已有的数据copy至新内存, O(n)复杂度.

 所以, Queue的enqueue操作很高效, 但dequeue操作有待优化:
 dequeue之后并不将剩下元素向前shift,而是将dequeue之后的空位留着,定期执行trim操作将所有空位移除即可.
 定期trim的复杂度为O(n), 但dequeue的平均复杂度降为O(1)了.
 因此,需要引入一个head,存储当前的实际首元素的index.
 */

let desc = ">>>>>>>>>> Queue采用Array实现, enqueue为O(1), dequeue为O(n). \n dequeue的优化方案:Array中可存储空值,并添加一个head存储当前实际首元素的index"

public struct Queue<T> {
    fileprivate var array = [T]()

    public var isEmpty: Bool {
        return array.isEmpty
    }

    public var count: Int {
        return array.count
    }

    public var top: T? {
        return array.first
    }

    public var bottom: T? {
        return array.last
    }

    public mutating func enqueue(_ newElement: T) {
        array.append(newElement)
    }

    public mutating func dequeue() -> T? {
        if isEmpty {
            return nil
        } else {
            return array.remove(at: 0)
        }
    }
}


public struct Queue_Optimized<T> {
    fileprivate var array = [T?]() // 元素可能为nil
    fileprivate var head = 0 // 真正的首元素所在的index

    public var isEmpty: Bool {
        return array.count - head == 0
    }

    public var count: Int {
        return array.count - head
    }

    public var top: T? {
        if array.count == 0 {
          return nil
        }
        return array[head]
    }

    public var bottom: T? {
      if array.count == 0 {
        return nil
      }
      return array.last!
    }

    public mutating func enqueue(_ newElement: T) {
        array.append(newElement)
    }

    public mutating func dequeue() -> T? {
        if array.count - head <= 0 {
          return nil
        }

        let headElement = array[head]
        array[head] = nil
        head = head + 1

        // remove nil elements
        let p = Double(head) / Double(array.count)
        if array.count > 100 && p > 0.25 {
            // removeFirst(_:) Removes the specified number of elements from the beginning of the collection.
            array.removeFirst(head)
            head = 0
        }

        if head == array.count {
            array.removeAll()
            head = 0
        }

        return headElement
    }
}


//////////////////////////////////////////////////

print(desc)


var queue_int = Queue<Int>()
queue_int.enqueue(1)
print(queue_int) // [1]
print(queue_int.isEmpty) // false
queue_int.enqueue(2)
print(queue_int) // [1,2]
queue_int.enqueue(3)
print(queue_int) // [1,2,3]
print("top : \(queue_int.top as Any)") // 1
print("bottom : \(queue_int.bottom as Any)") // 3
print("dequeue: \(queue_int.dequeue() as Any)") // 1
print(queue_int) // [2,3]
print("dequeue: \(queue_int.dequeue() as Any)") // 2
print(queue_int) // [3]
print("dequeue: \(queue_int.dequeue() as Any)") // 3
print(queue_int) // []
print(queue_int.isEmpty) // true

print("")

var queue_string = Queue<String>()
queue_string.enqueue("AAA")
print(queue_string) // ["AAA"]
print(queue_string.isEmpty) // false
queue_string.enqueue("BBB")
print(queue_string) // ["AAA", "BBB"]
queue_string.enqueue("CCC")
print(queue_string) // ["AAA", "BBB", "CCC"]
print("top : \(queue_string.top as Any)") // "AAA"
print("bottom : \(queue_string.bottom as Any)") // "CCC"
print("dequeue: \(queue_string.dequeue() as Any)") // "AAA"
print(queue_string) // ["BBB", "CCC"]
print("dequeue: \(queue_string.dequeue() as Any)") // "BBB"
print(queue_string) // ["CCC"]
print("dequeue: \(queue_string.dequeue() as Any)") // "CCC"
print(queue_string) // []
print(queue_string.isEmpty) // true

print("")

var queue_string_optimized = Queue_Optimized<String>()
queue_string_optimized.enqueue("AAA")
print(queue_string_optimized) // ["AAA"], head= 0
print(queue_string_optimized.isEmpty) // false
queue_string_optimized.enqueue("BBB")
print(queue_string_optimized) // ["AAA", "BBB"], head = 0
queue_string_optimized.enqueue("CCC")
print(queue_string_optimized) // ["AAA", "BBB", "CCC"], head = 0
print("top : \(queue_string_optimized.top as Any)") // "AAA"
print("bottom : \(queue_string_optimized.bottom as Any)") // "CCC"

print("dequeue: \(queue_string_optimized.dequeue() as Any)") // "AAA"
print(queue_string_optimized) // [nil, "BBB", "CCC"], head = 1
print("dequeue: \(queue_string_optimized.dequeue() as Any)") // "BBB"
print(queue_string_optimized) // [nil, nil, "CCC"], head = 2
print("dequeue: \(queue_string_optimized.dequeue() as Any)") // "CCC"
print(queue_string_optimized) // [], head = 0
print(queue_string_optimized.isEmpty) // true

print("top : \(queue_string_optimized.top as Any)") // nil
print("bottom : \(queue_string_optimized.bottom as Any)") // nil

File: js_algorithm/data_structure/queue.js
console.log(">>>>>>>>>> stack");

class Queue {
  constructor() {
    this.array = [];
  }

  isEmpty() {
    return this.array.length == 0;
  }

  count() {
    return this.array.length;
  }

  top() {
    return this.array[0];
  }

  bottom() {
    return this.array[this.array.length - 1];
  }

  enqueue(ele) {
    this.array.push(ele);
  }

  dequeue() {
    this.array.shift();
  }
}

// 引入head变量，存储当前实际首元素的index
// dequeue操作时并不将剩下元素向前shift，而是将dequeue的空位留着，定期执行trim操作清空空位即可
// trim操作的复制度为O(n)，但dequeue的平均复杂度降为O(1)。
class Queue_Optimized {
  constructor() {
    this.array = [];
    this.head = 0; // 实际有效的首元素所在的index
  }

  isEmpty() {
    return this.array.length - this.head == 0;
  }

  count() {
    return this.array.length - this.head;
  }

  top() {
    if (this.array.count == 0) {
      return null;
    }
    return this.array[this.head];
  }

  bottom() {
    if (this.array.count == 0) {
      return null;
    }
    return this.array[this.array.length - 1]
  }

  enqueue(ele) {
    this.array.push(ele);
  }

  dequeue(ele) {
    if (this.array.length - this.head <= 0) {
      return null
    }

    const headElement = this.array[this.head];
    this.array[this.head] = null,
    this.head += 1;

    if (this.head == this.array.length) {
      this.array = [];
      this.head = 0;
    }

    return headElement;
  }
}

var s = new Queue();
console.log(s.isEmpty()); // true
s.enqueue(1);
s.enqueue(2);
s.enqueue(3);
s.enqueue(4);
console.log(s); // [1,2,3,4]
s.dequeue();
console.log(s); // [2,3,4]
console.log(s.top()); // 2
console.log(s.bottom()); // 4
console.log(s.isEmpty()); // false

console.log("")

var so = new Queue_Optimized();
console.log(so.isEmpty()); // true
so.enqueue(1);
so.enqueue(2);
so.enqueue(3);
so.enqueue(4);
console.log(so); // [1,2,3,4], head = 0
so.dequeue();
console.log(so); // [null,2,3,4], head = 1
console.log(so.top()); // 2
console.log(so.bottom()); // 4
console.log(so.isEmpty()); // false
so.dequeue();
console.log(so); // [null,null,3,4], head = 2
so.dequeue();
console.log(so); // [null,null,null,4], head = 3
so.dequeue();
console.log(so); // [], head = 0
